src/test/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/TestsForJoinAlgorithms.java

```diff
@@ -0,0 +1,272 @@
+				assertEquals( "http://example.org/z1", b.get(var3).getURI() );
+				b1Found = true;
+			}
+			else if ( b.get(var1).getURI().equals("http://example.org/x2") ) {
+				assertEquals( "http://example.org/y2", b.get(var2).getURI() );
+				assertEquals( "http://example.org/z2", b.get(var3).getURI() );
+				b2Found = true;
+			}
+			else {
+				fail( "Unexpected URI for ?v1: " + b.get(var1).getURI() );
+			}
+		}
+
+		assertTrue(b1Found);
+		assertTrue(b2Found);
+	}
+
+	protected void _joinWithoutJoinVariable() {
+		final Var var1 = Var.alloc("v1");
+		final Var var2 = Var.alloc("v2");
+		final Var var3 = Var.alloc("v3");
+
+		final Node p = NodeFactory.createURI("http://example.org/p");
+		final Node x1 = NodeFactory.createURI("http://example.org/x1");
+		final Node x2 = NodeFactory.createURI("http://example.org/x2");
+		final Node y1 = NodeFactory.createURI("http://example.org/y1");
+		final Node y2 = NodeFactory.createURI("http://example.org/y2");
+		final Node z1 = NodeFactory.createURI("http://example.org/z1");
+		final Node z2 = NodeFactory.createURI("http://example.org/z2");
+
+		final GenericIntermediateResultBlockImpl input1 = new GenericIntermediateResultBlockImpl();
+		input1.add( SolutionMappingUtils.createSolutionMapping(var1, x1) );
+		input1.add( SolutionMappingUtils.createSolutionMapping(var1, x2) );
+
+		final GenericIntermediateResultBlockImpl input2 = new GenericIntermediateResultBlockImpl();
+		input2.add( SolutionMappingUtils.createSolutionMapping(
+				var2, y1,
+				var3, z1) );
+		input2.add( SolutionMappingUtils.createSolutionMapping(
+				var2, y2,
+				var3, z2) );
+
+		final Iterator<SolutionMapping> it = runTest(input1, input2);
+
+		assertTrue( it.hasNext() );
+		final Binding b1 = it.next().asJenaBinding();
+		assertEquals( 3, b1.size() );
+
+		assertTrue( it.hasNext() );
+		final Binding b2 = it.next().asJenaBinding();
+		assertEquals( 3, b2.size() );
+
+		assertTrue( it.hasNext() );
+		final Binding b3 = it.next().asJenaBinding();
+		assertEquals( 3, b3.size() );
+
+		assertTrue( it.hasNext() );
+		final Binding b4 = it.next().asJenaBinding();
+		assertEquals( 3, b4.size() );
+
+		assertFalse( it.hasNext() );
+	}
+
+	protected void _joinWithEmptyInput1() {
+		final Var var2 = Var.alloc("v2");
+		final Var var3 = Var.alloc("v3");
+
+		final Node p = NodeFactory.createURI("http://example.org/p");
+		final Node y1 = NodeFactory.createURI("http://example.org/y1");
+		final Node z1 = NodeFactory.createURI("http://example.org/z1");
+
+		final GenericIntermediateResultBlockImpl input1 = new GenericIntermediateResultBlockImpl();
+
+		final GenericIntermediateResultBlockImpl input2 = new GenericIntermediateResultBlockImpl();
+		input2.add( SolutionMappingUtils.createSolutionMapping(
+				var2, y1,
+				var3, z1) );
+
+		final Iterator<SolutionMapping> it = runTest(input1, input2);
+
+		assertFalse( it.hasNext() );
+	}
+
+	protected void _joinWithEmptyInput2() {
```

Comment:
Please add two more unit tests where one of the two inputs has one solution mapping which happens to be the empty solution mapping (which is compatible with any solution mapping).

Reply:
I added two more unit tests in TestsForJoinAlgorithms, please see commit f43e5b8

Reply:
Thanks for these two additional unit tests!